### PR TITLE
[fix] Nystrom + microGPT + some additive masking

### PR DESCRIPTION
**Pull request by blefaudeux**
## What does this PR do?
- Fixes https://github.com/facebookresearch/xformers/issues/74, microGPT and causal attention work properly with Nystrom
- Fix attention masks being bool or additive (not all the codepaths probably, but at least the ones that Nystrom touched)
- Fix the small-sequence handling for sparse masks, this was broken prior to this PR and probably explains the hanging CI yesterday

## Before submitting

- [x] Did you have fun?
  - Make sure you had fun coding 🙃
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [ ] Did you make sure to update the docs?
  - [ ] N/A
- [x] Did you write any new necessary tests?
  - [ ] N/A
- [x] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/master/CHANGELOG.md)? (if needed)
  - [ ] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

File: xformers/components/attention/core.py [hunk 1]
```diff
@@ -219,20 +223,31 @@ def scaled_dot_product_attention(
...
+    # Try to handle a case where the sequence is smaller than the mask
+    if (
+        att_mask is not None
+        and q.shape[-2] == k.shape[-2]
+        and q.shape[-2] < att_mask.shape[0]
+    ):
         seq = q.shape[-2]
-
-        if not att_mask.is_sparse:
-            att_mask = att_mask[:seq, :seq]
```
Comment (blefaudeux) [inline xformers/components/attention/core.py:227]: we were testing for smaller sequences, but when the mask was sparse it was not adjusted -> possible memory error which showed up on CI later on

**Comment by blefaudeux** [comment]
converting back to draft, I think that the key padding mask is not handled correctly with my changes

File: xformers/components/attention/nystrom.py [hunk 2]
```diff
@@ -163,32 +168,36 @@ def forward(
...
         if self.num_landmarks >= seq_len:
             mask: Optional[torch.Tensor] = None
             if self.causal:
-                mask = self._tril_mask(batched_dim, seq_len, seq_len)
+                mask = self._tril_mask(batched_dim, seq_len, seq_len, **tt)
             if key_padding_mask is not None:
-                mask = (
-                    key_padding_mask
-                    if mask is None
-                    else mask.logical_and(key_padding_mask)
```
Comment (blefaudeux) [inline xformers/components/attention/nystrom.py:190]: @dianaml0 I'm not sure of how that worked, since mask and key_padding_mask had different dimensions here, no ?

Comment (dianaml0) [reply xformers/components/attention/nystrom.py:190]: It should automatically broadcast key_padding_mask along the mismatched dimension, similar to https://github.com/pytorch/pytorch/blob/4262c8913c2bddb8d91565888b4871790301faba/torch/nn/functional.py#L5189

File: xformers/components/attention/nystrom.py [hunk 3]
```diff
@@ -161,34 +166,41 @@ def forward(
...
 
         if key_padding_mask is not None:
-            assert key_padding_mask.dtype == torch.bool
+            if key_padding_mask.dtype == torch.bool:
+                logging.warning(
+                    "Bool mask found, but an additive mask is expected. Converting but this is slow"
+                )
+                key_padding_mask = bool_mask_to_additive(key_padding_mask)
+
+            assert key_padding_mask is not None  # mypy is drunk
```
Comment (dianaml0) [inline xformers/components/attention/nystrom.py:183]: I think if a return type is added to `bool_mask_to_additive` it may fix the mypy error

**Review by dianaml0 — APPROVED** [review]
Thanks!! Really helpful changes! Makes masking easier to work with!

File: xformers/components/attention/nystrom.py (hunk 3 above)
Comment (blefaudeux) [reply xformers/components/attention/nystrom.py:183]: ahh good point, I'll fix that, thank you !
